src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs

```diff
@@ -250,73 +256,129 @@ protected override IntPtr LoadUnmanagedDll(string unmanagedDllName)
 
             public BasicClassFactory(Guid clsid, Type classType)
             {
-                this.classId = clsid;
-                this.classType = classType;
+                _classId = clsid;
+                _classType = classType;
             }
 
-            public void CreateInstance(
-                [MarshalAs(UnmanagedType.Interface)] object pUnkOuter,
-                ref Guid riid,
-                [MarshalAs(UnmanagedType.Interface)] out object ppvObject)
+            public static void ValidateInterfaceRequest(Type classType, ref Guid riid, object outer)
             {
-                if (riid != Marshal.IID_IUnknown)
+                Debug.Assert(classType != null);
+                if (riid == Marshal.IID_IUnknown)
                 {
-                    bool found = false;
+                    return;
+                }
 
-                    // Verify the class implements the desired interface
-                    foreach (Type i in this.classType.GetInterfaces())
-                    {
-                        if (i.GUID == riid)
-                        {
-                            found = true;
-                            break;
-                        }
-                    }
+                // Aggregation can only be done when requesting IUnknown.
+                if (outer != null)
+                {
+                    const int CLASS_E_NOAGGREGATION = unchecked((int)0x80040110);
+                    throw new COMException(string.Empty, CLASS_E_NOAGGREGATION);
+                }
+
+                bool found = false;
 
-                    if (!found)
+                // Verify the class implements the desired interface
+                foreach (Type i in classType.GetInterfaces())
+                {
+                    if (i.GUID == riid)
                     {
-                        // E_NOINTERFACE
-                        throw new InvalidCastException();
+                        found = true;
+                        break;
                     }
                 }
 
-                ppvObject = Activator.CreateInstance(this.classType);
+                if (!found)
+                {
+                    // E_NOINTERFACE
+                    throw new InvalidCastException();
+                }
+            }
+
+            public static object CreateAggregatedObject(object pUnkOuter, object comObject)
+            {
+                Debug.Assert(pUnkOuter != null && comObject != null);
+                try
+                {
+                    IntPtr outerPtr = Marshal.GetIUnknownForObject(pUnkOuter);
+                    IntPtr innerPtr = Marshal.CreateAggregatedObject(outerPtr, comObject);
+                    return Marshal.GetObjectForIUnknown(innerPtr);
+                }
+                finally
+                {
+                    // Decrement the above 'Marshal.GetIUnknownForObject()'
+                    Marshal.ReleaseComObject(pUnkOuter);
```

Comment:
If `GetIUnknownForObject` fails, this will still decrement. Should `GetIUnknownForObject` be outside the `try` ?